### ISSUE TITLE
CA2217 False positive if setting value of enum using shift operator

### ISSUE THREAD

**Issue by appel1** [dotnet/roslyn-analyzers#5086]
### Analyzer

**Diagnostic ID**: [CA2217](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2217): `Do not mark enums with FlagsAttribute`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.3 (Latest)

### Describe the bug

False positive

### Steps To Reproduce

Analyse code like this:

```csharp
[Flags]
    public enum Something
    {
        None = 0,
        First = 1 << 0,
        Second = 1 << 1,
        Third = 1 << 2,
        Fourth = 1 << 3,
        Last = 1 << 4,
    }
```

### Expected behavior

No violation

### Actual behavior

Violation: 
error CA2217: Do not mark enums with FlagsAttribute

**Comment by NewellClark** [comment]
I've added unit tests to detect this bug and am unable to get it to repro. [Here](https://github.com/NewellClark/roslyn-analyzers/tree/%235086-false-positive-CA2217) is the branch I've been working on.

**Comment by appel1** [comment]
It is possible that I was a bit too quick with my steps to reproduce. 

You need at least 32 non-zero values.

```csharp
[Flags]
public enum Something
{
    None = 0,
    One = 1 << 0,
    Two = 1 << 1,
    Three = 1 << 2,
    Four = 1 << 3,
    Five = 1 << 4,
    Six = 1 << 5,
    Seven = 1 << 6,
    Eight = 1 << 7,
    Nine = 1 << 8,
    Ten = 1 << 9,
    Eleven = 1 << 10,
    Twelve = 1 << 11,
    Thirteen = 1 << 12,
    Fourteen = 1 << 13,
    Fifteen = 1 << 14,
    Sixteen = 1 << 15,
    Seventeen = 1 << 16,
    Eighteen = 1 << 17,
    Nineteen = 1 << 18,
    Twenty = 1 << 19,
    TwentyOne = 1 << 20,
    TwentyTwo = 1 << 21,
    TwentyThree = 1 << 22,
    TwentyFour = 1 << 23,
    TwentyFive = 1 << 24,
    TwentySix = 1 << 25,
    TwentySeven = 1 << 26,
    TwentyEight = 1 << 27,
    TwentyNine = 1 << 28,
    Thirty = 1 << 29,
    ThirtyOne = 1 << 30,
    ThirtyTwo = 1 << 31,
}
```

![image](https://user-images.githubusercontent.com/973314/118311806-aba80c80-b4f0-11eb-9024-22f37410d68b.png)

**Comment by NewellClark** [comment]
@appel1 Here's what's happening: when you say `1 << 31`, you get -2147483648, because `1` is an `int` constant. The analyzer is correctly telling you that you have an invalid flags enum, because flags enumerations are supposed to have all even powers of 2.

**Comment by appel1** [comment]
@NewellClark Don't I feel silly now. Good thing NetAnalzyers caught this in our code then, it'll be interesting to find out if this was done intentionally or not. 🤔

**Comment by mavasani** [comment]
Closing as by design.
